contrib/lkt/language/parser.py

```diff
@@ -470,6 +470,31 @@ def first_no_paren_parent():
             Entity.parent
         )
 
+    @langkit_property(public=True, return_type=T.LKNode.entity)
+    def topmost_invalid_decl(
+        topmost=(T.LKNode.entity, No(T.LKNode.entity))
+    ):
+        """
+        Return the ``topmost`` FullDecl annotated with ``@invalid``,
+        null otherwise. ``topmost`` argument is the topmost declaration
+        annotated with ``@invalid`` from the current node to the root node if
+        it exists, null otherwise. It is used to get the first ``@invalid``
+        declaration in case where they are nested.
```

Comment:
Actually, it does start from `Entity` (and therefore can return it): it starts by looking at the parent in order to start the recursion from it, by setting `topmost` to `Entity` if it is an `@invalid` declaration. There is only one case where it cannot return `Entity`: when it has no parent, i.e. `Entity` is the root node, which is OK since it cannot be a declaration annotated with `@invalid`.
```
        return Entity.parent.then(
            lambda p: p.topmost_invalid_decl(
                Entity.is_a(FullDecl).then(
                    lambda _: Entity.cast(FullDecl)
                    .has_annotation('invalid').then(
                        lambda _: Entity,             # <- starts from Entity
                        default_val=topmost
                    ),
                    default_val=topmost
                )
            ),
            default_val=topmost
        )
```

Reply:
Oh I see, you are correct, thanks. I find this specific recursion hard to proof-read, there are odd repetitions (`default_val=topmost` appears 3 times) and the fact that the topmost argument (which is there only as an implementation detail) leaks to the user API is a smell. I would suggest to use `.parents()` in order to switch to an iterative (and straightforward) form. Something such as:

```python
return Self.parents().filter(
    lambda p:
    p.cast(FullDecl).then(lambda fd: fd.has_annotation("invalid"))
).at(-1)
```

Reply:
Done!